**Grupo:** 1 <br>
**Integrantes:**
* Jerson Alexander Peña Martínez
* Alejandro López-Fando Fernández
* Manuela Larrea Gómez


### **Trabajo en grupo: Primera entrega**
<hr>

### Requerimientos del sistema

* Versión de Python: 3.11.4 o superior
* Versión del módulo Pandas: 2.0.1 o superior

# **Consumo de drogas: Atributos de la personalidad relacionados** 




### **1. Comprender el problema  y establecer objetivos** 

### **2. Set de datos objetivo**

El dataset utilizado fue obtenido de la siguiente dirección: https://archive.ics.uci.edu/dataset/373/drug+consumption+quantified

#### 2.1 Data card: (hacer tabla con HTML o pegar imagen)

#### 2.2 Importación de librerías

In [63]:
import pandas as pd

#### 2.3 Importación de los datos

In [64]:
df = pd.read_csv('../data/drug_consumption.data', header = None)
df.head(5)                 

0        1        2        3        4        5        6        7        8   \
0   1  0.49788  0.48246 -0.05921  0.96082  0.12600  0.31287 -0.57545 -0.58331   
1   2 -0.07854 -0.48246  1.98437  0.96082 -0.31685 -0.67825  1.93886  1.43533   
2   3  0.49788 -0.48246 -0.05921  0.96082 -0.31685 -0.46725  0.80523 -0.84732   
3   4 -0.95197  0.48246  1.16365  0.96082 -0.31685 -0.14882 -0.80615 -0.01928   
4   5  0.49788  0.48246  1.98437  0.96082 -0.31685  0.73545 -1.63340 -0.45174   

        9   ...   22   23   24   25   26   27   28   29   30   31  
0 -0.91699  ...  CL0  CL0  CL0  CL0  CL0  CL0  CL0  CL2  CL0  CL0  
1  0.76096  ...  CL4  CL0  CL2  CL0  CL2  CL3  CL0  CL4  CL0  CL0  
2 -1.62090  ...  CL0  CL0  CL0  CL0  CL0  CL0  CL1  CL0  CL0  CL0  
3  0.59042  ...  CL0  CL0  CL2  CL0  CL0  CL0  CL0  CL2  CL0  CL0  
4 -0.30172  ...  CL1  CL0  CL0  CL1  CL0  CL0  CL2  CL2  CL0  CL0  

[5 rows x 32 columns]

El set de datos no dispone de encabezado, se procede a renombrar cada una de las columnas de acuerdo con la información proporcionada en la documentación de la fuente: UC Irvine Machine Learning Repository (https://archive.ics.uci.edu/dataset/373/drug+consumption+quantified)

In [65]:
df.columns = ['ID','Age (Real)','Gender (Real)','Education (Real)','Country (Real)',
              'Ethnicity (Real)','Nscore (Real)','Escore (Real)','Oscore (Real)','Ascore (Real)',
              'Cscore (Real)','Impulsive (Real)','SS (Real)','Alcohol','Amphet','Amyl','Benzos',
              'Caff','Cannabis','Choc','Coke','Crack','Ecstasy','Heroin','Ketamine','Legalh','LSD',
              'Meth','Mushrooms','Nicotine','Semer','VSA']
df.head(10)

ID  Age (Real)  Gender (Real)  Education (Real)  Country (Real)  \
0   1     0.49788        0.48246          -0.05921         0.96082   
1   2    -0.07854       -0.48246           1.98437         0.96082   
2   3     0.49788       -0.48246          -0.05921         0.96082   
3   4    -0.95197        0.48246           1.16365         0.96082   
4   5     0.49788        0.48246           1.98437         0.96082   
5   6     2.59171        0.48246          -1.22751         0.24923   
6   7     1.09449       -0.48246           1.16365        -0.57009   
7   8     0.49788       -0.48246          -1.73790         0.96082   
8   9     0.49788        0.48246          -0.05921         0.24923   
9  10     1.82213       -0.48246           1.16365         0.96082   

   Ethnicity (Real)  Nscore (Real)  Escore (Real)  Oscore (Real)  \
0           0.12600        0.31287       -0.57545       -0.58331   
1          -0.31685       -0.67825        1.93886        1.43533   
2          -0.31685       -0.46725        0.80523       -0.84732   
3          -0.31685       -0.14882       -0.80615       -0.01928   
4          -0.31685        0.73545       -1.63340       -0.45174   
5          -0.31685       -0.67825       -0.30033       -1.55521   
6          -0.31685       -0.46725       -1.09207       -0.45174   
7          -0.31685       -1.32828        1.93886       -0.84732   
8          -0.31685        0.62967        2.57309       -0.97631   
9          -0.31685       -0.24649        0.00332       -1.42424   

   Ascore (Real)  ...  Ecstasy  Heroin  Ketamine Legalh  LSD Meth Mushrooms  \
0       -0.91699  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   
1        0.76096  ...      CL4     CL0       CL2    CL0  CL2  CL3       CL0   
2       -1.62090  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL1   
3        0.59042  ...      CL0     CL0       CL2    CL0  CL0  CL0       CL0   
4       -0.30172  ...      CL1     CL0       CL0    CL1  CL0  CL0       CL2   
5        2.03972  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   
6       -0.30172  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   
7       -0.30172  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   
8        0.76096  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   
9        0.59042  ...      CL0     CL0       CL0    CL0  CL0  CL0       CL0   

  Nicotine Semer  VSA  
0      CL2   CL0  CL0  
1      CL4   CL0  CL0  
2      CL0   CL0  CL0  
3      CL2   CL0  CL0  
4      CL2   CL0  CL0  
5      CL6   CL0  CL0  
6      CL6   CL0  CL0  
7      CL0   CL0  CL0  
8      CL6   CL0  CL0  
9      CL6   CL0  CL0  

[10 rows x 32 columns]

### **3. Estadísticas descriptivas preliminares**

In [66]:
df.shape

(1885, 32)

In [67]:
df.dtypes

ID                    int64
Age (Real)          float64
Gender (Real)       float64
Education (Real)    float64
Country (Real)      float64
Ethnicity (Real)    float64
Nscore (Real)       float64
Escore (Real)       float64
Oscore (Real)       float64
Ascore (Real)       float64
Cscore (Real)       float64
Impulsive (Real)    float64
SS (Real)           float64
Alcohol              object
Amphet               object
Amyl                 object
Benzos               object
Caff                 object
Cannabis             object
Choc                 object
Coke                 object
Crack                object
Ecstasy              object
Heroin               object
Ketamine             object
Legalh               object
LSD                  object
Meth                 object
Mushrooms            object
Nicotine             object
Semer                object
VSA                  object
dtype: object

In [68]:
df.describe()

ID  Age (Real)  Gender (Real)  Education (Real)  \
count  1885.000000  1885.00000    1885.000000       1885.000000   
mean    945.294960     0.03461      -0.000256         -0.003806   
std     545.167641     0.87836       0.482588          0.950078   
min       1.000000    -0.95197      -0.482460         -2.435910   
25%     474.000000    -0.95197      -0.482460         -0.611130   
50%     946.000000    -0.07854      -0.482460         -0.059210   
75%    1417.000000     0.49788       0.482460          0.454680   
max    1888.000000     2.59171       0.482460          1.984370   

       Country (Real)  Ethnicity (Real)  Nscore (Real)  Escore (Real)  \
count     1885.000000       1885.000000    1885.000000    1885.000000   
mean         0.355542         -0.309577       0.000047      -0.000163   
std          0.700335          0.166226       0.998106       0.997448   
min         -0.570090         -1.107020      -3.464360      -3.273930   
25%         -0.570090         -0.316850      -0.678250      -0.695090   
50%          0.960820         -0.316850       0.042570       0.003320   
75%          0.960820         -0.316850       0.629670       0.637790   
max          0.960820          1.907250       3.273930       3.273930   

       Oscore (Real)  Ascore (Real)  Cscore (Real)  Impulsive (Real)  \
count    1885.000000    1885.000000    1885.000000       1885.000000   
mean       -0.000534      -0.000245      -0.000386          0.007216   
std         0.996229       0.997440       0.997523          0.954435   
min        -3.273930      -3.464360      -3.464360         -2.555240   
25%        -0.717270      -0.606330      -0.652530         -0.711260   
50%        -0.019280      -0.017290      -0.006650         -0.217120   
75%         0.723300       0.760960       0.584890          0.529750   
max         2.901610       3.464360       3.464360          2.901610   

         SS (Real)  
count  1885.000000  
mean     -0.003292  
std       0.963701  
min      -2.078480  
25%      -0.525930  
50%       0.079870  
75%       0.765400  
max       1.921730

In [69]:
# El dataset tiene muchas variables y no es posible visualizar de forma cómoda la data. 
# Se opta por visualizar el DataFrame transpuesto para 5 observaciones al azar

df.sample(5).T

1668     624      1543     1805     1693
ID                   1672      627     1547     1809     1697
Age (Real)        1.82213  0.49788 -0.95197 -0.95197 -0.95197
Gender (Real)    -0.48246  0.48246 -0.48246  0.48246 -0.48246
Education (Real)  0.45468 -0.05921 -1.22751  1.16365 -0.61113
Country (Real)    0.96082  0.96082 -0.57009  0.96082 -0.57009
Ethnicity (Real) -0.31685 -0.31685   0.1144 -1.10702 -0.31685
Nscore (Real)     0.62967 -0.34799 -0.24649  0.41667  0.13606
Escore (Real)    -0.43999 -0.57545 -1.92173 -0.80615  -1.7625
Oscore (Real)     1.24033 -0.58331  0.29338 -0.58331 -0.45174
Ascore (Real)     0.94156 -0.91699 -0.60633  0.13136  0.28783
Cscore (Real)    -1.13788  0.25953 -1.13788 -0.40581 -0.40581
Impulsive (Real) -0.71126 -0.71126  0.52975 -1.37983 -0.71126
SS (Real)        -0.21575 -1.54858  0.40148 -1.54858 -0.52593
Alcohol               CL5      CL6      CL3      CL4      CL3
Amphet                CL0      CL1      CL0      CL0      CL0
Amyl                  CL0      CL0      CL0      CL0      CL0
Benzos                CL0      CL0      CL0      CL0      CL5
Caff                  CL6      CL6      CL6      CL5      CL5
Cannabis              CL1      CL2      CL4      CL0      CL4
Choc                  CL6      CL6      CL4      CL5      CL5
Coke                  CL0      CL2      CL0      CL0      CL0
Crack                 CL0      CL0      CL0      CL0      CL0
Ecstasy               CL0      CL2      CL4      CL0      CL0
Heroin                CL0      CL0      CL0      CL0      CL0
Ketamine              CL0      CL0      CL0      CL0      CL0
Legalh                CL0      CL0      CL3      CL0      CL0
LSD                   CL0      CL2      CL4      CL0      CL0
Meth                  CL0      CL0      CL0      CL0      CL6
Mushrooms             CL0      CL2      CL3      CL0      CL0
Nicotine              CL1      CL2      CL2      CL0      CL0
Semer                 CL0      CL0      CL0      CL0      CL0
VSA                   CL0      CL0      CL0      CL0      CL0

Se constata que el dataset dispone de 1885 observaciones, es decir, 1885 encuentados de acuerdo con el datacard de la fuente. <br>
Cada observación dispone de 32 atributos, de las cuales 19 de tipo object y el resto son de tipo float64 e int64. <br>
Acorde con la información de la fuente y lo hallado en esta etapa preliminar, ninguna de las variables son númericas. Todas las variables son categóricas y 12 de ellas han sido cuantificadas (aquellas de tipo float64).  


### **4. Preprocesamiento y limpieza de los datos**

El preprocesamiento y limpieza tiene como objetivo mejorar la calidad de los datos y los resultados de la minería. Esta sección incluye: <br>
* Eliminación de ruido y datos aislados o outliers.
* Eliminar inconsistencias y duplicados.
* Imputar información faltante.
 

#### 4.1 Preprocesamiento de datos
* Incompletos: atributos que carecen de valores, atributos sin interés (missing values).
* Ruidosos: contienen errores o “outliers”
* Inconsistentes: discrepancias en códigos o nombres.
* Tamaño excesivo: filas y/o columnas.

In [70]:
# Existencia de valores null 
df.isnull().values.any()

False

In [71]:
# Reducción tamaño del dataframe: Número de columnas
'''
La variable "ID" no ofrece información valiosa para el análisis planteado en el objetivo del proyecto, debido a que es el número
de registro en la base de datos original (a la cual no tenemos acceso). Además, acorde con la fuente, este ID no puede estar relacionado 
con el participante. 

Por lo anterior, se opta por eliminarla. 

'''
df = df.drop(labels=['ID'], axis=1) 

In [72]:
# Verificación de inconsistencias: Discrepacias en los datos respecto a los códigos o nombres definidos por la fuente (data card).

def crear_diccionarios_valores_unicos(data):
  """
  Crea un diccionario por cada columna del conjunto de datos, compuesto por sus valores únicos

  Args:
    data: El conjunto de datos

  Returns:
    Un diccionario de diccionarios, donde la clave es el nombre del atributo y el valor es un conjunto con los valores únicos
  """

  diccionarios = {}
  for atributo in data.columns:
    valores_unicos = set(data[atributo].unique())
    diccionarios[atributo] = valores_unicos
  return diccionarios

In [85]:
diccionarios = crear_diccionarios_valores_unicos(df)
diccionarios

{'Age (Real)': {-0.95197, -0.07854, 0.49788, 1.09449, 1.82213, 2.59171},
 'Gender (Real)': {-0.48246, 0.48246},
 'Education (Real)': {-2.43591,
  -1.7379,
  -1.43719,
  -1.22751,
  -0.61113,
  -0.05921,
  0.45468,
  1.16365,
  1.98437},
 'Country (Real)': {-0.57009,
  -0.46841,
  -0.28519,
  -0.09765,
  0.21128,
  0.24923,
  0.96082},
 'Ethnicity (Real)': {-1.10702,
  -0.50212,
  -0.31685,
  -0.22166,
  0.1144,
  0.126,
  1.90725},
 'Nscore (Real)': {-3.46436,
  -3.15735,
  -2.75696,
  -2.52197,
  -2.42317,
  -2.3436,
  -2.21844,
  -2.05048,
  -1.86962,
  -1.69163,
  -1.55078,
  -1.43907,
  -1.32828,
  -1.1943,
  -1.05308,
  -0.92104,
  -0.79151,
  -0.67825,
  -0.58016,
  -0.46725,
  -0.34799,
  -0.24649,
  -0.14882,
  -0.05188,
  0.04257,
  0.13606,
  0.22393,
  0.31287,
  0.41667,
  0.52135,
  0.62967,
  0.73545,
  0.82562,
  0.91093,
  1.02119,
  1.13281,
  1.23461,
  1.37297,
  1.49158,
  1.60383,
  1.72012,
  1.8399,
  1.98437,
  2.127,
  2.28554,
  2.46262,
  2.61139,
  2.82196,


In [82]:
for atributo, valores_unicos in diccionarios.items():
    print("Atributo:", atributo)
    print("Número de valores únicos:", len(valores_unicos))

Atributo: Age (Real)
Número de valores únicos: 6
Atributo: Gender (Real)
Número de valores únicos: 2
Atributo: Education (Real)
Número de valores únicos: 9
Atributo: Country (Real)
Número de valores únicos: 7
Atributo: Ethnicity (Real)
Número de valores únicos: 7
Atributo: Nscore (Real)
Número de valores únicos: 49
Atributo: Escore (Real)
Número de valores únicos: 42
Atributo: Oscore (Real)
Número de valores únicos: 35
Atributo: Ascore (Real)
Número de valores únicos: 41
Atributo: Cscore (Real)
Número de valores únicos: 41
Atributo: Impulsive (Real)
Número de valores únicos: 10
Atributo: SS (Real)
Número de valores únicos: 11
Atributo: Alcohol
Número de valores únicos: 7
Atributo: Amphet
Número de valores únicos: 7
Atributo: Amyl
Número de valores únicos: 7
Atributo: Benzos
Número de valores únicos: 7
Atributo: Caff
Número de valores únicos: 7
Atributo: Cannabis
Número de valores únicos: 7
Atributo: Choc
Número de valores únicos: 7
Atributo: Coke
Número de valores únicos: 7
Atributo: C

No hay observaciones con valores únicos por atributo, que no se encuentren pre definidos en el data card de la fuente. 

### **6. Análisis de los datos**

1. ¿Qué factores influyen en el consumo de sustancias? El dataset incluye información sobre factores demográficos, psicológicos y conductuales que podrían estar relacionados con el consumo de sustancias. Por ejemplo, se podría analizar si hay diferencias en el consumo de sustancias según la edad, el género, el nivel de educación o la etnicidad.

2. ¿Qué tipos de sustancias son más consumidas por las personas con diferentes características? El dataset incluye información sobre el consumo de una amplia gama de sustancias, desde alcohol y tabaco hasta drogas ilegales. Se podría analizar si hay diferencias en el consumo de estas sustancias según las características de las personas.

3. Cuales son las drogas mayormente consumidas? Hacer agrupaciones por las variables independientes. 

4. Cuáles son los fármacos con alta correlación? Esta correlacion estara dada por mivel de consumo, estas correlaciones permiten establecer patrones de consumo? 

5. Cuáles son las sustancias con baja correlación? 

6. Cuales son las caracteristicas comunes entre los grupos de consumidores y no consumidores? 

7. ¿Cómo se relacionan el consumo de sustancias y la personalidad? El dataset incluye información sobre la personalidad de las personas, evaluada mediante cuestionarios psicológicos. Se podría analizar si hay relaciones entre el consumo de sustancias y las diferentes dimensiones de la personalidad.

8. ¿Cómo se relacionan el consumo de sustancias y el comportamiento? El dataset incluye información sobre el comportamiento de las personas, como su estado civil, su ocupación o su nivel de ingresos. Se podría analizar si hay relaciones entre el consumo de sustancias y el comportamiento de las personas.
